File: go/1466.go
package main

type P struct {
	n  int
	in bool
}

func minReorder(n int, connections [][]int) (ans int) {
	g := make([][]P, n)
	for _, conn := range connections {
		f, t := conn[0], conn[1]
		g[f] = append(g[f], P{t, false})
		g[t] = append(g[t], P{f, true})
	}
	var dfs func(root, fa int)
	dfs = func(root, fa int) {
		for _, p := range g[root] {
			if p.n != fa {
				if !p.in {
					ans++
				}
				dfs(p.n, root)
			}
		}
	}
	dfs(0, -1)
	return
}

File: python/1466.py
from typing import List

class Solution:
    def minReorder(self, n: int, connections: List[List[int]]) -> int:
        g = [[] for _ in range(n)]
        for f, t in connections:
            g[f].append([t, False])
            g[t].append([f, True])
        ans = 0
        def dfs(root: int, fa: int) -> None:
            nonlocal ans
            for x in g[root]:
                if x[0] != fa:
                    if not x[1]:
                        ans += 1
                    dfs(x[0], root)
        dfs(0, -1)
        return ans